### ISSUE TITLE
Alerting: Migrations run in an unexpected order

### ISSUE THREAD

**Issue by alexweav** [grafana/grafana#60196]
Our database migrations are not running in the order that they were written. I'm not aware of any existing issues around this, but I'm shocked this hasn't caused anything yet.

This is because of how the migration code is structured. Migrations are broken up by table, so there is not one linear history.

A contrived example of how it currently looks:
```
AddMigrations() {
    AddTableXMigrations()
    AddTableYMigrations()
}

AddTableXMigrations() {
    MigrationOne()
    MigrationThree()
}

AddTableYMigrations() {
    MigrationTwo()
}
```
In this example, the migrations run in this order: 1->3->2. This is the wrong order.

We've largely been safe so far by pure chance, as most migrations are very simplistic and are scoped to only a single table. The instant we add a migration which touches multiple tables (i.e. needs to run in a specific point in time relative to two tables, not one), this becomes extremely bug-prone and leads to upgrade issues as table schemas drift.

In the example above, assume a migration touches both tables X and Y. **There is no bug-free place to put this migration.**
- If it's put in `AddMigrations()`, it breaks unless it's defined at the very end, after all table-specific migrations. Then, any future migration added to `AddTableXMigrations` or `AddTableYMigrations` then gets incorrectly injected _before_ our new migration, breaking it if it changes the schema.
- If it's put in `AddTableXMigrations()`, table Y won't exist yet, and the migration instantly fails.
- If it's put in `AddTableYMigrations()`, a future migration defined on table X again ends up again incorrectly running too early, breaking the schema out from underneath us.

Solving this is simple. We instead define all migrations in a single, append-only, linear history, regardless of table:

```
AddMigrations() {
    MigrationOne()
    MigrationTwo()
    MigrationThree()
}
```

**Comment by joeblubaugh** [comment]
We can probably use `git blame` to get a candidate ordering. Probably still need to verify the order is safe manually though.

**Comment by alexweav** [comment]
I believe this should be done in two steps:

First, I think we should extract all migrations into a linear history in the order that they currently execute today. Since this change doesn't touch the order of any existing migrations, we can have high confidence in it - and, any future migrations are made safer.

Then, we can look into re-ordering historical migrations. Joe's suggestion of `git blame` above sounds like a great approach to me.

**Comment by JacobsonMT** [comment]
@alexweav Do you have an example of where we have an `AddTableXMigrations` function (or similiar) where we add multiple migrations in the manner mentioned above? I'm having trouble figuring out where our issues are specifically.

**Comment by alexweav** [comment]
It can break basically any "non-trivial" migration. I'm defining non-trivial as something that touches multiple tables or is not backwards compatible (i.e. renaming a column).

I just added one - the moving of records to the alert_configuration_history table is an example of this.

Basically, if anyone makes a change to the `alert_configuration` table or the `alert_configuration_history` table in tables.go like we have been, it will actually run *before* the migration I added, and cause it to break. That one is a specific example.

Still, the issue is more widespread - migrations are very dependent on the order they're run. It's concerning that when we write a migration, it's not being run in the context we think. For trivial migrations, this can be safe by way of luck, but it's the kind of thing that's a major bug just waiting to happen.

**Comment by JacobsonMT** [comment]
Ah, so it has more to do with the structuring of our table migrations being one table at a time. The intention should have been to make sure we don't modify any existing migration but presenting it like `AddTableXMigrations` implies that new ones should go in there instead of at the end.

This is not something we can really prevent with better structuring but better structuring would definitely lessen then risk of a developer mistakenly modifying an existing migration instead of correctly appending to the end.

As you said, if we assume the current order of migrations is OK (even if it's not historically correct), we can put these aside with a big "DO NOT EDIT" notice and go forward from here with some better naming. I don't know what benefit we'll really gain from trying to reorder the existing migrations to be historically correct, in fact I think it's risk without much reward.

If we're not ok with just using exorbitant warnings, there are probably ways to prevent modifications in pre-commit/pre-process/CI, however, I think this might be overkill.

**Comment by alexweav** [comment]
I 100% agree, @JacobsonMT . Let's restructure to prevent issues in the future, but we can probably safely assume that the current ones are okay, since we haven't had many complex migrations beyond the rollout of unified alerting yet.

**Comment by JacobsonMT** [comment]
@alexweav I created a PR that tries to improve the comments and segments off the existing table migrations into a separate function. I also moved your new configuration history migrations into `tables.go` and could move some others as well to maybe keep `ualert` specific to the main alerting migration that is not immutable.

Let me know if this is what you had in mind or if there's more you had intended.
